### ISSUE TITLE
StepFunctions StateMachine cannot yet  be updated by Cloudformation

### ISSUE THREAD

**Issue by prestomation** [spulec/moto#3402]
In my project, we use CDK to generate a large cloudformation template
We feed this cloudformation template into moto in our pytest suite so a number of things 'just work' like they would in the cloud.

I'm trying to upgrade to 1.3.15, and this setup broke.
When referencing a Statemachine in CDK, CDK emits cloudformation like this
`"Fn::Select": [6, {"Fn::Split": [":", {"Ref": "MyCoolStateMachine"}]}]}, "*"]]}}`

This ends up giving us this error

```
resource_json = {'Fn::Split': [':', {'Ref': 'MyCoolStateMachine'}]}, resources_map = <moto.cloudformation.parsing.ResourceMap object at 0x7fd2728eb220>

    def clean_json(resource_json, resources_map):
        """
        Cleanup the a resource dict. For now, this just means replacing any Ref node
        with the corresponding physical_resource_id.

        Eventually, this is where we would add things like function parsing (fn::)
        """
        if isinstance(resource_json, dict):
            if "Ref" in resource_json:
                # Parse resource reference
                resource = resources_map[resource_json["Ref"]]
                if hasattr(resource, "physical_resource_id"):
                    return resource.physical_resource_id
                else:
                    return resource

            if "Fn::FindInMap" in resource_json:
                map_name = resource_json["Fn::FindInMap"][0]
                map_path = resource_json["Fn::FindInMap"][1:]
                result = resources_map[map_name]
                for path in map_path:
                    if "Fn::Transform" in result:
                        result = resources_map[clean_json(path, resources_map)]
                    else:
                        result = result[clean_json(path, resources_map)]
                return result

            if "Fn::GetAtt" in resource_json:
                resource = resources_map.get(resource_json["Fn::GetAtt"][0])
                if resource is None:
                    return resource_json
                try:
                    return resource.get_cfn_attribute(resource_json["Fn::GetAtt"][1])
                except NotImplementedError as n:
                    logger.warning(str(n).format(resource_json["Fn::GetAtt"][0]))
                except UnformattedGetAttTemplateException:
                    raise ValidationError(
                        "Bad Request",
                        UnformattedGetAttTemplateException.description.format(
                            resource_json["Fn::GetAtt"][0], resource_json["Fn::GetAtt"][1]
                        ),
                    )

            if "Fn::If" in resource_json:
                condition_name, true_value, false_value = resource_json["Fn::If"]
                if resources_map.lazy_condition_map[condition_name]:
                    return clean_json(true_value, resources_map)
                else:
                    return clean_json(false_value, resources_map)

            if "Fn::Join" in resource_json:
                join_list = clean_json(resource_json["Fn::Join"][1], resources_map)
                return resource_json["Fn::Join"][0].join([str(x) for x in join_list]) 

            if "Fn::Split" in resource_json:
                to_split = clean_json(resource_json["Fn::Split"][1], resources_map)
>               return to_split.split(resource_json["Fn::Split"][0])

```

Looks like the resource Map value is `None`, if I had it guess it's because StateMachine doesn't implement CloudformationModel.

We are working around this problem with a simple monkey patch 
```

    from moto.cloudformation import parsing 
    cj = parsing.clean_json
    
    def new_cj(json, map):
        cleaned = cj(json, map)
        if cleaned is None:
            if json["Ref"].endswith("StateMachine"):
                return f'arn:aws:states:us-east-1:111122223333:stateMachine:{json["Ref"]}'
        return cleaned
    parsing.clean_json = new_cj
```

This is clearly not ideal, but allows us to upgrade. 

I was not able to find another issue for SFN support in Cloudformation

**Comment by prestomation** [comment]
Similarly, when deleting the stack above we see

```
    def delete(self):
        remaining_resources = set(self.resources)
        tries = 1
        while remaining_resources and tries < 5:
            for resource in remaining_resources.copy():                parsed_resource = self._parsed_resources.get(resource)
                try:
                    if parsed_resource and hasattr(parsed_resource, "delete"):
                        parsed_resource.delete(self._region_name)
                    else:
                        if hasattr(parsed_resource, "physical_resource_id"):
                            resource_name = parsed_resource.physical_resource_id
                        else:
                            resource_name = None

                        resource_json = self._resource_json_map[
>                           parsed_resource.logical_resource_id
                        ]

```
Which is likely the Statemachine that has a `None` resource.

**Comment by bblommers** [comment]
Thanks for raising this, @prestomation. 
In 1.3.14, there was no support for SFN at all, so I guess it just ignored that setup. We did introduce basic support for it in 1.3.15, but indeed no CloudFromation support yet. Marking it as an enhancement.

FYI: 1.3.15 has a bug surrounding dependency management, so you might be better off upgrading to 1.3.16 - literraly the only change in that version is a bugfix for 1.3.15.

**Comment by bblommers** [comment]
CloudFormation creation and deletion of a StateMachine is now available as of moto >= 1.3.16.dev89.

Will keep this open as a placeholder to integrate CF updates.

**Comment by bblommers** [comment]
CloudFormation updates for StateMachines are now available as of moto >=1.3.16.dev96
